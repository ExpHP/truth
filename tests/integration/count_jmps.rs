#[allow(unused)]
use crate::integration_impl::{expected, formats::*, Format};

const FORMAT_WITH_GT_JMP: Format = ECL_08;
const FORMAT_WITH_NE_JMP: Format = ANM_12;

source_test!(
    FORMAT_WITH_NE_JMP, ne_jmp_can_roundtrip,
    main_body: r#"
        label:
            if (--I2) goto label;
    "#,
    check_decompiled: |decompiled| { assert!(decompiled.contains("(--$REG[10002])")); },
);

source_test!(
    FORMAT_WITH_GT_JMP, gt_jmp_can_roundtrip,
    main_body: r#"
        label:
            if (--I2 > 0) goto label;
    "#,
    check_decompiled: |decompiled| { assert!(decompiled.contains("(--$REG[10002] > 0)")); },
);

source_test!(
    FORMAT_WITH_GT_JMP, ne_jmp_can_suggest_gt,
    main_body: r#"
        label:
            if (--I2) goto label;  //~ ERROR (--I2 > 0)
    "#,
);

source_test!(
    FORMAT_WITH_NE_JMP, gt_jmp_can_suggest_ne,
    main_body: r#"
        label:
            if (--I2 > 0) goto label;  //~ ERROR (--I2)
    "#,
);

source_test!(
    FORMAT_WITH_NE_JMP, ne_format_can_compile_times,
    main_body: r#"
        times(I2=5) { nop(); }
    "#,
    check_compiled: |_, _| { /* just compile */ },
);

source_test!(
    FORMAT_WITH_GT_JMP, gt_format_can_compile_times,
    main_body: r#"
        times(I2=5) { nop(); }
    "#,
    check_compiled: |_, _| { /* just compile */ },
);
