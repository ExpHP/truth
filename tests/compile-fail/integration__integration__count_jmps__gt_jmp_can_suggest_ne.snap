---
source: tests/integration/count_jmps.rs
expression: stderr
---
error: feature not supported by format
   ┌─ <input>:29:13
   │
29 │             if (--I2 > 0) goto label;
   │             ^^^^^^^^^^^^^^^^^^^^^^^^^ decrement jump of this form not supported in this game
   │
   = this language supports a different form of decrement jump; try 'if (--I2)'


